### ISSUE TITLE
C++14

### ISSUE THREAD

**Issue by marthtz** [eclipse/iceoryx#220]
## Brief feature description
As described in CONTRIBUTING.MD we'd like to move on to support C++ 14.

## Detailed information
What's holding us back?

**Comment by elBoberido** [comment]
At the moment, I think QCC 5.4 on QNX. But we might soon switch to QCC8. Then there would be nothing which prevents us to switch to C++14
🤞

**Comment by mossmaurice** [comment]
> At the moment, I think QCC 5.4 on QNX. But we might soon switch to QCC8. Then there would be nothing which prevents us to switch to C++14
> 🤞

AFAIK QCC 5.4 already fully supports C++14. As it is based on GCC see https://gcc.gnu.org/projects/cxx-status.html#cxx14

**Comment by ithier** [comment]
Hard agree.
Gimme `auto` in lambdas :P (and other goodies).

**Comment by elBoberido** [comment]
I'm the last one who wouldn't want to use C++14, but currently it seems we get ASIL D QCC 5.4 only for C++11.

There are discussions to switch to QCC 8, where we would get C++14.
In the end we also need to switch to C++14 if we want to be compliant with the AUTOSAR Adaptive platform.
@budrus, @dkroenke, do you have more information about the compiler we need to support? Should we just switch to C++14?

**Comment by elBoberido** [comment]
It seems QNX 7.1 with QCC 8.3 was released about 2 weeks ago
https://blackberry.qnx.com/en/embedded-software/qnx-software-development-platform

**Comment by MiaoDX** [comment]
Hi, it seems that you are moving on to C++14, as one possible downstream user, our project will still stick to C++ in the foreseeable future. So, if the API/ABI/linking is compatible, we won't complain, if not, please consider making ver 1.0 C++11 -.-

Discussions on stackoverflow seems that I should not worry too much:
[Discuss 1](https://stackoverflow.com/questions/32787244/can-i-use-c14-in-a-library-meant-for-c11-client-applications) and [Discuss 2](https://stackoverflow.com/questions/46746878/is-it-safe-to-link-c17-c14-and-c11-objects)

Looking forward to 1.0 ~

**Comment by dejanpan** [comment]
@elBoberido replying to https://github.com/eclipse/iceoryx/issues/220#issuecomment-669055820:
1. I am not aware of any issues with the gcc compiler on QNX
1. The most recent versions of QNX had these versions of gcc and libc++
    1. QNX SDP 7.0 => GCC 5.4 with libcxx 3.7.1
    1. QNX SDP 7.1 = > GCC 8.3 with libcxx 8.0

If we check the status of gcc and libc++ we indeed see that all C++14 features are implemented in the lowest of the above versions:
1. https://gcc.gnu.org/projects/cxx-status.html#cxx14
1. https://libcxx.llvm.org/cxx1y_status.html 

QNX also has a certified version (QNX OS for Safety 2.2) https://backendnews.net/blackberry-qnx-adds-another-certification-to-its-embedded-software-portfolio/ but this one is also OK wrt the C++14.

**Comment by elBoberido** [comment]
@dejanpan totally agree
@MiaoDX sorry for not replying that long. It somehow went under the radar. Additionally to what dejapan said, we are also introducing a C-API, so the absolute worst case would be to use that one instead of the C++ API
@marthtz @budrus any objections from your side to switch in the near future, let's say in November?

**Comment by dejanpan** [comment]
The discussion that is slightly harder is whether `iceoryx` should switch to C++17.

For that I would like to give a perspective from the point of view of our company, Apex.AI, that is certifying ROS 2 (currently still written in C++ 14) according to ISO 26262 (ASIL-D).

For this discussion relevant are these simplified steps that happen on the code:
1. Code is compiled with additional security and safety related compiler flags (e.g. `-Werror=format-security`, `-Wcast-align`, `-pedantic`, `-Wunused`, ...)
1. Code is analysed with the [Axivion](https://www.axivion.com/en) static code analyser using [Adaptive Autosar Coding Guidelines](https://www.autosar.org/fileadmin/user_upload/standards/adaptive/19-11/AUTOSAR_RS_CPP14Guidelines.pdf)
1. Code is analysed with the [Vector Cast](https://www.vector.com/us/en-us/products/products-a-z/software/vectorcast/?gclid=EAIaIQobChMIzYmL1oCd5wIVC8RkCh1GkQT8EAAYASAAEgKeV_D_BwE#c74806) tool and the test coverage is computed for lines, functions, branches and [MC/DC](https://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/20010057789.pdf)
1. Architecture and the dependency structure of the code is analysed with [Lattix Architect](https://www.lattix.com/products-architecture-issues/)
1. Control and data flows are analysed with the tool [Understand](https://www.scitools.com/features/)
1. Test stubs are (auto) created with VectorCast for the dependent units (system calls, ...)
1. Unit tests are written in `gtest`
1. The code is sanitized with the various sanitizers: https://clang.llvm.org/docs/index.html
1. The code is cross-compiled for Embedded Linux and QNX on `aarch64` architectures
    1. In the near-term future other POSIX-compliant [RTOSes](https://en.wikibooks.org/wiki/Embedded_Systems/Common_RTOS) will be added
1. We use the `gcc` and `libc++` versions available on above RTOSes and which have been validated with the validation suites like this one: https://solidsands.com/supertest.
1. Following back-ported features from C++17 are used: std::optional, std::variant, std::string_view, std::apply, polymorphic allocator
1. The code undergoes performance testing using https://gitlab.com/ApexAI/performance_test
1. The performance is analysed and tuned with the following tools: `perf`, `heaptrack`, `valgrind`, `LTT-ng`

---

For us to now decide whether we can switch to the newest version of C++ we need to consider if above RTOSes and tools support, in this case, C++17.
1. QNX is OK because of what [I said above](https://github.com/eclipse/iceoryx/issues/220#issuecomment-724923069) and:
    1. https://gcc.gnu.org/projects/cxx-status.html#cxx17
    1. https://libcxx.llvm.org/cxx1z_status.html 
1. VxWorks also already listed [C++17 support](https://www.windriver.com/news/press/pr.html?ID=22444)
1. VectorCAST and Axivion both use the C++ front end from EDG (https://www.edg.com/c)  and they stay current with the latest standards. EDG fully supports C++17 already, and work is underway to support C++20

---

What is unknown/show stopper:
1. Adaptive Autosar Coding Guidelines for C++17 are not available yet
1. Someone will have to create a toolchain file for the cross compilation on the ARM-based ECUs
1. C++17 support in these tools: Lattix, Understand, LTT-ng (likely available)
1. We do not know what is the willingness of the (automotive) customers to switch to C++17. So far everyone is set on C++14.

**In conclusion:** Apex.AI would be willing to switch to C++17 as soon as the Adaptive Autosar Coding Guidelines for C++17 will be available.

**Comment by marthtz** [comment]
>
>QNX also has a certified version (QNX OS for Safety 2.2) https://backendnews.net/blackberry-qnx-adds-another-certification-to-its-embedded-software-portfolio/ but this one is also OK wrt the C++14.

Sounds good. QNX SDP 7 mentions ISO 26262 ASIL D for automotive and C++14 support. However, the "Revision List of the QNX OS for Safety (QOS)" lists the C++ Library as ASIL B?
https://www.certipedia.com/fs-products/files/certificates/certificates_asi/2020/EZ/968_EZ_653_13_20/appendix/EZ_653_13_20_RL_2020-11-02_QOS_2_2.pdf

**Comment by dejanpan** [comment]
@marthtz that is correct. Which means that for ASIL-D applications that are using such ASIL B certified libc++ one will need to certify the used parts of the libc++ as in context of that application.
